### PR TITLE
Fix schedule reset logic in pytorch profiler

### LINKED ISSUE #10717: Pytorch Profiler causes memory leak
## 🐛 Bug

It seems like chosing the Pytorch profiler causes an ever growing amount of RAM being allocated. This even continues after training, probably while the profiler data is processed.

After a certain number of epochs, this causes an OOM and triggers my Kernel to kill the process.

### To Reproduce

To reproduce, simply enable the profiler on one of the provided examples

```sh
cd pl_examples/basic_examples/mnist_examples
python image_classifier_5_lightning_datamodule.py --trainer.profiler=pytorch --trainer.gpus=1
```

On my machine, sometime mid epoch=3, I am OOM and the process gets killed.

### Expected behavior

The memory leak does not occur

### Environment

```
 CUDA:
        - GPU:
                - NVIDIA GeForce GTX 1060 6GB
        - available:         True
        - version:           10.2
* Packages:
        - numpy:             1.21.4
        - pyTorch_debug:     False
        - pyTorch_version:   1.10.0+cu102
        - pytorch-lightning: 1.6.0dev
        - tqdm:              4.62.3
* System:
        - OS:                Linux
        - architecture:
                - 64bit
                - ELF
        - processor:         
        - python:            3.8.12
        - version:           #1 ZEN SMP PREEMPT Thu, 18 Nov 2021 22:23:53 +0000
```

### Additional context

I am aware that this might be caused by Pytorch and not Lightning and I am currently trying to reproduce this issue in plain Pytorch. If I can reproduce it, this issue can of course be triaged to them.

cc @tchaton @carmocca @kaushikb11 @ninginthecloud

### PR DESCRIPTION
**Pull request by rohitgr7**
## What does this PR do?

Fixes #10717

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

cc @borda @tchaton @carmocca @kaushikb11 @ninginthecloud

### CODE REVIEW

File: pytorch_lightning/profiler/pytorch.py [hunk 1]
```diff
@@ -335,9 +335,24 @@ def _init_kineto(self, profiler_kwargs: Any) -> None:
...
+        if self._schedule._current_action == "predict_step":
+            return sum(trainer.num_predict_batches)
+
     def _should_override_schedule(self) -> bool:
-        return (self._lightning_module is not None and self._lightning_module.trainer.limit_train_batches < 5) and (
-            self._schedule is not None and self._schedule._schedule == self._default_schedule()
+        return (
+            self._lightning_module is not None
+            and self._schedule is not None
+            and self._total_steps < 5
```
Comment (rohitgr7) [inline pytorch_lightning/profiler/pytorch.py:354]: actual fix is this.

others are mypy issues

**Review by tchaton — APPROVED** [review]
LGTM !

**Review by ethanwharris — APPROVED** [review]

File: pytorch_lightning/loops/dataloader/prediction_loop.py [hunk 2]
```diff
@@ -53,10 +53,7 @@ def num_dataloaders(self) -> int:
     @property
     def max_batches(self) -> List[int]:
         """The max number of batches this loop will run for each dataloader."""
-        max_batches = self.trainer.num_predict_batches
-        if isinstance(max_batches, int):
-            max_batches = [max_batches] * len(self.dataloaders)
```
Comment (awaelchli) [inline pytorch_lightning/loops/dataloader/prediction_loop.py:58]: curious, are you removing this due to a redundancy, or because it is part of the fix? 
I suppose there is already logic in data_loading.py that takes care of this, so `isinstance(max_batches, int)` is never True?

**Review by awaelchli — APPROVED** [review]

File: pytorch_lightning/loops/dataloader/prediction_loop.py (hunk 2 above)
Comment (rohitgr7) [reply pytorch_lightning/loops/dataloader/prediction_loop.py:58]: not a part of the fix, but it's redundant. was trying to fix some mypy issues and found this. num_pred_batches is always a list of ints.
